shared/app/nav.desktop.js

```diff
@@ -1,46 +1,40 @@
 // @flow
+import * as I from 'immutable'
 import * as React from 'react'
+import * as RouteTree from '../route-tree/render-route'
 import GlobalError from './global-errors/container'
-import Offline from '../offline'
+import Offline from '../offline/container'
 import TabBar from './tab-bar/container'
 import {Box, ErrorBoundary} from '../common-adapters'
-import {chatTab, loginTab, peopleTab, profileTab, type Tab} from '../constants/tabs'
-import {connect, type TypedState} from '../util/container'
+import {chatTab, loginTab, type Tab} from '../constants/tabs'
+import {connect, type TypedState, type Dispatch} from '../util/container'
 import {globalStyles} from '../styles'
-import {navigateTo, switchTo} from '../actions/route-tree'
-import {createShowUserProfile} from '../actions/profile-gen'
-import {type Props, type StateProps, type DispatchProps, type OwnProps} from './nav'
 
-class Nav extends React.Component<Props> {
-  _switchTab = tab => {
-    this.props.switchTab(tab)
-  }
+type Props = {
+  layerScreens: I.Stack<RouteTree.RenderRouteResult>,
+  visibleScreen: RouteTree.RenderRouteResult,
+  routeSelected: Tab,
+  routePath: I.List<string>,
+}
 
+class Nav extends React.Component<Props> {
   render() {
-    const props = this.props
-    const visibleScreen = props.routeStack.findLast(r => !r.tags.layerOnTop)
-    if (!visibleScreen) {
-      throw new Error('no route component to render without layerOnTop tag')
-    }
-    const layerScreens = props.routeStack.filter(r => r.tags.layerOnTop)
+    const {routeSelected, routePath, visibleScreen, layerScreens} = this.props
     return (
       <ErrorBoundary>
         <Box style={stylesTabsContainer}>
-          {props.routeSelected !== loginTab && (
-            <TabBar onTabClick={this._switchTab} selectedTab={props.routeSelected} />
-          )}
+          {routeSelected !== loginTab && <TabBar routeSelected={routeSelected} routePath={routePath} />}
           <ErrorBoundary>
             <Box style={{...globalStyles.flexBoxColumn, flex: 1}}>
-              {visibleScreen.component({isActiveRoute: true, shouldRender: true})}
-              {layerScreens.map(r => r.leafComponent({isActiveRoute: true, shouldRender: true}))}
+              {/* We use a fixed key here so we don't remount components like chat. */}
+              {visibleScreen.component({key: '0', shouldRender: true})}
```

Comment:
before the route tree used the path as a key but in cases (like in chat) where the path is parameterized for the same view hiearchy this causes a massive amount of render thrashing. aka clicking inbox items would cause almost the entire tree to unmount / remount with the exact same components